akka-actor-tests/src/test/scala/akka/actor/CoordinatedShutdownSpec.scala

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.actor
+
+import scala.concurrent.Await
+import scala.concurrent.Future
+
+import akka.Done
+import akka.testkit.AkkaSpec
+import com.typesafe.config.ConfigFactory
+
+object CoordinatedShutdownSpec {
+
+}
+
+class CoordinatedShutdownSpec extends AkkaSpec {
+
+  def extSys = system.asInstanceOf[ExtendedActorSystem]
+
+  "CoordinatedShutdown" must {
+
+    "sort phases in topolgical order" in {
+      CoordinatedShutdown.topologicalSort(Map.empty) should ===(Nil)
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "a" → Set.empty)) should ===(List("a"))
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "b" → Set("a"))) should ===(List("a", "b"))
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "c" → Set("a"), "b" → Set("a"))).head should ===("a") // b, c can be in any order
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "b" → Set("a"), "c" → Set("b"))) should ===(List("a", "b", "c"))
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "b" → Set("a"), "c" → Set("a", "b"))) should ===(List("a", "b", "c"))
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "c" → Set("a", "b"))).last should ===("c") // a, b can be in any order
+
+      CoordinatedShutdown.topologicalSort(Map(
+        "b" → Set("a"), "c" → Set("b"), "d" → Set("b", "c"), "e" → Set("d"))) should ===(
+        List("a", "b", "c", "d", "e"))
+
+      val result = CoordinatedShutdown.topologicalSort(Map(
+        "a2" → Set("a1"), "a3" → Set("a2"),
+        "b2" → Set("b1"), "b3" → Set("b2")))
+      val (a, b) = result.partition(_.charAt(0) == 'a')
+      a should ===(List("a1", "a2", "a3"))
+      b should ===(List("b1", "b2", "b3"))
+    }
+
+    "detect cycles in phases (non-DAG)" in {
+      intercept[IllegalArgumentException] {
+        CoordinatedShutdown.topologicalSort(Map(
+          "b" → Set("a"), "a" → Set("b")))
+      }
+
+      intercept[IllegalArgumentException] {
+        CoordinatedShutdown.topologicalSort(Map(
+          "c" → Set("a"), "c" → Set("b"), "b" → Set("c")))
+      }
+
+      intercept[IllegalArgumentException] {
+        CoordinatedShutdown.topologicalSort(Map(
+          "d" → Set("a"), "d" → Set("c"), "c" → Set("b"), "b" → Set("d")))
+      }
+
+    }
```

Comment:
Also, just for completeness sake, please add a test for a self-cycle, i.e. `a` depends on `a`